File: Restoration/11_UnityReskin/Assets/Plugins/Android/CoinMergeHaptics.androidlib/src/com/coinmerge/recovery/NativeHaptics.java
package com.coinmerge.recovery;

import android.content.Context;
import android.os.Build;
import android.os.VibrationEffect;
import android.os.Vibrator;

public final class NativeHaptics {
    private NativeHaptics() {}
    public static void pulse(Context context, int milliseconds, int amplitude) {
        if (context == null) return;
        Vibrator vibrator = (Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE);
        if (vibrator == null || !vibrator.hasVibrator()) return;
        int duration = Math.max(1, milliseconds);
        if (Build.VERSION.SDK_INT >= 26) {
            vibrator.vibrate(VibrationEffect.createOneShot(duration, Math.max(1, Math.min(255, amplitude))));
        } else {
            vibrator.vibrate(duration);
        }
    }
}

File: Restoration/11_UnityReskin/Assets/Scripts/Gameplay/NativeHaptics.cs
using System;
using UnityEngine;
namespace CoinMerge.Recovery
{
    // Explicit JNI method binding, not System.Reflection/Java reflection.
    public static class NativeHaptics
    {
        static IntPtr bridge,method,unityPlayer,activityField;
        static readonly jvalue[] arguments=new jvalue[3];
        public static void Pulse(int milliseconds,int amplitude)
        {
            if(Application.platform==RuntimePlatform.Android)
            {
                if(bridge==IntPtr.Zero)
                {
                    var local=AndroidJNI.FindClass("com/coinmerge/recovery/NativeHaptics");
                    bridge=AndroidJNI.NewGlobalRef(local);AndroidJNI.DeleteLocalRef(local);
                    method=AndroidJNI.GetStaticMethodID(bridge,"pulse","(Landroid/content/Context;II)V");
                    local=AndroidJNI.FindClass("com/unity3d/player/UnityPlayer");
                    unityPlayer=AndroidJNI.NewGlobalRef(local);AndroidJNI.DeleteLocalRef(local);
                    activityField=AndroidJNI.GetStaticFieldID(unityPlayer,"currentActivity","Landroid/app/Activity;");
                }
                var activity=AndroidJNI.GetStaticObjectField(unityPlayer,activityField);
                arguments[0].l=activity;arguments[1].i=milliseconds;arguments[2].i=amplitude;
                try{AndroidJNI.CallStaticVoidMethod(bridge,method,arguments);}finally{AndroidJNI.DeleteLocalRef(activity);}
            }
            else if(Application.platform==RuntimePlatform.IPhonePlayer)Handheld.Vibrate();
        }
    }
}
